test/rubygems/test_gem_package_tar_reader_entry.rb

```diff
@@ -211,4 +257,67 @@ def test_zero_byte_file_readpartial
   ensure
     close_util_entry(zero_entry) if zero_entry
   end
+
+  def test_read_from_gzip_io
+    tgz = util_gzip(@tar)
+
+    Zlib::GzipReader.wrap StringIO.new(tgz) do |gzio|
+      header = Gem::Package::TarHeader.from gzio
+      entry = Gem::Package::TarReader::Entry.new header, gzio
+
+      assert_equal @contents, entry.read
+      entry.rewind
+      assert_equal @contents, entry.read, "second read after rewind should read same contents"
+    end
+  end
+
+  def test_read_from_gzip_io_with_non_zero_offset
+    skip = @tar.size
+    contents2 = ("0".."9").to_a.join * 100
+    @tar << tar_file_header("lib/bar", "", 0, contents2.size, Time.now)
+    @tar << contents2
+    @tar << "\0" * (512 - (@tar.size % 512))
+
+    tgz = util_gzip(@tar)
+
+    Zlib::GzipReader.wrap StringIO.new(tgz) do |gzio|
+      gzio.read(skip) # skip the first entry so io.pos is not 0, preventing easy rewind
+      header = Gem::Package::TarHeader.from gzio
+      entry = Gem::Package::TarReader::Entry.new header, gzio
+
+      assert_equal contents2, entry.read
+      entry.rewind
+      assert_equal contents2, entry.read, "second read after rewind should read same contents"
+    end
+  end
+
+  def test_seek_in_gzip_io_with_non_zero_offset
+    skip = @tar.size
+    contents2 = ("0".."9").to_a.join * 100
+    @tar << tar_file_header("lib/bar", "", 0, contents2.size, Time.now)
+    @tar << contents2
+    @tar << "\0" * (512 - (@tar.size % 512))
```

Comment:
Saw this logic in a few places, we could move it to a `util_tar_contents` method or something like that.

Reply:
Updated. Looks a lot better. Thanks!